Implemented torch.inner

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46716 Implemented torch.inner**

Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

heitorschueroff committed:
Implemented torch.inner

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [ ] Add documentation

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [ ] Add documentation

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [ ] Add documentation

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,70 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
+  // Broadcast all but the last dimension
+  Tensor input = self;
+  for (int64_t i = other.dim() - 1; i > 0; --i) {
+    input = input.unsqueeze(-2);
```
`input.unsqueeze(-2)`? That seems odd.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,70 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
+  // Broadcast all but the last dimension
+  Tensor input = self;
+  for (int64_t i = other.dim() - 1; i > 0; --i) {
+    input = input.unsqueeze(-2);
```
Same question here as in the `out` implementation. How does this unsqueeze happening multiple times relate to broadcasting?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3113,6 +3113,49 @@ def merge_dicts(*dicts):
+add_docstr(torch.inner, r"""
+inner(input, other, *, out=None) -> Tensor
+
+This functions computes the inner product of :attr:`input` and :attr:`other`
```
What's an inner product? Follow-up question, what's a sum-product?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3113,6 +3113,49 @@ def merge_dicts(*dicts):
+
+.. note:: 
+
+    If either input is a scalar, then a elemnt-wise multiplication is performed.
```
If :attr:\`input\` or :attr:\`other\` is a scalar then this is equivalent to `torch.mul(a, b)`. If both :attr:\`input\` and :attr:\`other\` are tensors then their last (innermost) dimensions must have the same size.

mruberry commented on test/test_linalg.py:
```diff
@@ -19,6 +19,44 @@
     exact_dtype = True
 
+    @dtypes(torch.float, torch.double)
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
We now require NumPy to run the test suite so these checks are no longer needed.

mruberry commented on test/test_linalg.py:
```diff
@@ -19,6 +19,44 @@
 class TestLinalg(TestCase):
     exact_dtype = True
 
+    @dtypes(torch.float, torch.double)
```
torch.double variant is probably not very interesting; I think this can just test torch.float?

mruberry commented on test/test_linalg.py:
```diff
@@ -19,6 +19,44 @@
+        check([2], [3, 0, 2])       # 1D x 3D empty
+
+        check([1, 2], [3, 2])       # 2D x 2D
+        check([1, 2], [3, 4, 2])    # 2D x 3D
```
Add a 4D x 4D test case, too

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,72 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
+
+  // out_shape = self_shape[:-1] + other_shape[:-1]
+  std::vector<int64_t> shape(
+      self.sizes().begin(), self.sizes().begin() + self.dim() - 1);
```
begin->cbegin

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,72 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
+      self.sizes().begin(), self.sizes().begin() + self.dim() - 1);
+  shape.insert(
+      shape.end(),
+      other.sizes().begin(),
```
begin->cbegin

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,72 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
 }
 
+Tensor& inner_out(Tensor& out, const Tensor& self, const Tensor& other) {
+  checkDeviceType("inner()", {out, self, other}, self.device().type());
```
Wait a minute... isn't `inner` = `np.tensordot(a, b, axes=(-1,-1))` => `torch.tensordot(a, b, dims=((-1,), (-1,)))`?

heitorschueroff replied:
torch.inner(a, b) == torch.tensordot(a, b, dims=([-1], [-1])) for non-scalar a and b only. For scalars, it behaves differently.

mruberry replied:
OK, so we're halfway there. What happens for scalars? Can we write the op as a composite of tensordot plus some scalar handling?

heitorschueroff replied:
Actually we can, torch.tensordot(a, b, dims=0) is equivalent to torch.inner if either a or b are scalars.

heitorschueroff replied:
On second thought, why replace a multiplication by a call to tensordot? Same number of lines but worse performance.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3113,6 +3113,56 @@ def merge_dicts(*dicts):
+add_docstr(torch.inner, r"""
+inner(input, other, *, out=None) -> Tensor
+
+Computes the vector inner product for 1D tensors. In higher dimensions, computes the sum
```
Going to delay reviewing docs for a moment until we resolve the equivalency to `torch.tensordot(a_t, b_t, dims=((-1,), (-1,)))`, since if that's the case the docs become easier to write.

heitorschueroff replied:
Replied above.

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -186,6 +186,47 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
+
+  // If either self or other is a scalar just multiply them
+  if (self.dim() == 0 || other.dim() == 0) {
+    resize_output(out, self.dim() == 0 ? other.sizes() : self.sizes());
```
The mul_out will handle the resize for you, right?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3160,6 +3160,53 @@ def merge_dicts(*dicts):
+inner(input, other, *, out=None) -> Tensor
+
+Computes the vector inner product for 1D tensors. For higher dimensions, computes
+a contraction of :attr:`input` and :attr:`other` over their last dimension.
```
I think this documentation is very good and we can probably ship it as-is. Maybe in the future we can think more about terms like "contraction"?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3160,6 +3160,53 @@ def merge_dicts(*dicts):
+add_docstr(torch.inner, r"""
+inner(input, other, *, out=None) -> Tensor
+
+Computes the vector inner product for 1D tensors. For higher dimensions, computes
```
vector inner product -> dot product?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3160,6 +3160,53 @@ def merge_dicts(*dicts):
+
+Example::
+
+    # Vector inner product
```
If above change is "dot product" need to make this consistent

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3160,6 +3160,53 @@ def merge_dicts(*dicts):
+    tensor(7)
+
+    # Multidimensional input tensors
+    >>> torch.inner(torch.randn(2, 3), torch.randn(2, 4, 3))
```
Examples using randn are tricky because the original values are unknown. Maybe use arange or construct the tensors explicitly?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3160,6 +3160,53 @@ def merge_dicts(*dicts):
+
+    # Scalar input
+    >>> a = torch.randn(2, 3)
+    >>> a
```
This use of randn is OK because it shows the tensor's values

mruberry commented on test/test_linalg.py:
```diff
@@ -19,6 +19,51 @@
+        check([2, 1, 3, 2], [1, 3, 2, 2])   # 4D x 4D
+
+        # Test complex numbers
+        a = torch.tensor([2 + 1j, 3 - 2j])
```
Why not just add cfloat as a dtype?

mruberry commented on test/test_linalg.py:
```diff
@@ -19,6 +19,51 @@
+        self.assertEqual(a.inner(b).cpu().numpy(), np.inner(a.cpu().numpy(), b.cpu().numpy()))
+
+        # Test discontiguous input
+        a = torch.randn(3, 2, device=device, dtype=dtype).transpose_(0, 1)
```
Metanote: the test suite should have a tool that makes it easy to construct discontiguous tensors from a shape (i.e. `randn_discontiguous(...)`)

heitorschueroff replied:
Where is this? I tried doing a global search but didn't find any use of it.

mruberry replied:
It doesn't exist yet, which is why I was just making a note about it. Sorry that was confusing.

mruberry reviewed (approved):
This is a really "clean" PR. I made a final few comments/suggestions that should be easy tweaks.

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

[ghstack-poisoned]

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -1521,6 +1521,12 @@ def add_docstr_all(method, docstr):
 See :func:`torch.ger`
 """)
 
+add_docstr_all('inner', r"""
```
Update tensors.rst, too, so this appears

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3173,6 +3173,64 @@ def merge_dicts(*dicts):
 
 """)
 
+add_docstr(torch.inner, r"""
```
Update torch.rst, too, so this appears

mruberry commented:
I think if you rebase the errors will go away. Just remember to update the .rst files.

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch.inner"


Implemented torch.inner similar to [numpy.inner](https://numpy.org/doc/stable/reference/generated/numpy.inner.html). For now it's implemented as a composite op.

TODO

- [x] Add documentation

Differential Revision: [D24860351](https://our.internmc.facebook.com/intern/diff/D24860351)

[ghstack-poisoned]
